### PR TITLE
Implement minimal TCP module ("net")

### PR DESCRIPTION
**Pull request by matiasinsaurralde**
Similar to #209 

I've added tests and also a sample echo server in `samples/echoserver.ts`.

### CODE REVIEW

File: msg.proto [hunk 1]
```diff
@@ -99,4 +108,11 @@ message Msg {
   bytes write_file_sync_data = 131;
   uint32 write_file_sync_perm = 132;
   // write_file_sync_perm specified by https://godoc.org/os#FileMode
+
+  int32 net_port = 134;
```
Comment (jedahan) [inline msg.proto]: What is the reason this is 134 instead of 133?

File: net.ts [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+export function Socket(): NetSocket {
+    const socket = new NetSocket();
+    return socket;
+}
+
+export class NetServerConn {
+    private readonly id: number;
+    private onDataCb: (data: Uint8Array) => void;
+    constructor(id: number) {
```
[the remark this answers is not in the document]
Comment (matiasinsaurralde) [reply net.ts]: 👍

File: msg.proto (hunk 1 above)
Comment (matiasinsaurralde) [reply msg.proto]: Fixed

File: deno.d.ts [hunk 3]
```diff
@@ -11,4 +12,38 @@ declare module "deno" {
     data: Uint8Array,
     perm: number
   ): void;
-}
+
+  export class NetSocket {
+    private connectCb: () => void;
```
Comment (yordis) [inline deno.d.ts]: `onConnect` maybe?

File: deno.d.ts [hunk 4]
```diff
@@ -11,4 +12,38 @@ declare module "deno" {
...
+    private onDataCb: (data: Uint8Array) => void;
```
Comment (yordis) [inline deno.d.ts]: `onData` without the `Cb`. It reads the same.

File: net.go [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+	client, ok := clients[msg.NetClientId]
+	if !ok {
+	}
+	client.Write(msg.NetData)
+}
+
+func netClientClose(msg *Msg) {
+	client, ok := clients[msg.NetClientId]
+	if !ok {
+	}
```
Comment (gillchristian) [inline net.go]: Why is this block empty? Is it just WIP? (same for `netClientWrite` above)

File: net.go [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+	PubMsg(netChan, okMsg)
+}
+
+func netSocketRead(id int32, conn net.Conn) {
+	reader := bufio.NewReader(conn)
+	for {
+		data, _, err := reader.ReadLine()
+		if err != nil {
+			return
```
Comment (gillchristian) [inline net.go]: On [`netSocketRead`](https://github.com/matiasinsaurralde/deno/blob/2b59bf260313628eb3b0b39ebd4fd94b960e71ea/net.go#L85-L99) you `break` on error when reading line. Any reason for a `return` here, shouldn't it also be `break` here?

File: tests.ts [hunk 7]
```diff
@@ -55,3 +57,30 @@ test(async function tests_writeFileSync() {
   const actual = dec.decode(dataRead);
   assertEqual("Hello", actual);
 });
+
+test(function test_createServer() {
+  const decoder = new TextDecoder("utf-8");
+  const server = createServer((conn: NetServerConn) => {
+    // console.log("conn writes");
+    // conn.write("welcome\n");
```
Comment (gillchristian) [inline tests.ts]: Maybe remove this commented code?

**Comment by matiasinsaurralde** [comment]
@gillchristian Thanks for the comments, just updated the code

**Comment by ry** [comment]
@matiasinsaurralde Awesome!! Looks quite good. I will review in more detail soon. (Very sorry for the delays.)

**Comment by ry** [comment]
@matiasinsaurralde Let's land your http patch (#209 ) first. Many of my comments in that PR will apply here.

**Comment by matiasinsaurralde** [comment]
Just FYI: have refactored the messages a bit, might still need tweaks.
Also added the `wg` calls for proper blocking when the server runs.
Added the `close` method in both TS and Go but the implementation is missing, I've added a `TODO` comment, could be for another PR.

**Comment by ry** [comment]
@matiasinsaurralde We will come back to this soon but I'm closing this PR for now.

**Pull request closed without merging**